### PR TITLE
Add major vscale modules and tests

### PR DESCRIPTION
**Pull request by meetshah1995**

### CODE REVIEW

File: riscv/PC_mux.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    :param Signal PC_PIF: Program counter IF current
+    """
+
+    base = Signal(modbv(0)[XPR_LEN:])
+    offset = Signal(modbv(0)[XPR_LEN:])
+
+    @always_comb
+    def PC_mux_assign():
+
+        imm_b = modbv(0)[XPR_LEN:]
```
Comment (jck) [inline riscv/PC_mux.py]: Why modbv?

File: riscv/csr_file.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+                elif addr == CSR_ADDR_INSTRETHW:
+                    instret_full.next[2 * XPR_LEN:XPR_LEN] = wdata_internal
+                elif addr == CSR_ADDR_TO_HOST:
+                    to_host.next = wdata_internal
+                elif addr == CSR_ADDR_FROM_HOST:
+                    from_host.next = wdata_internal
+            if htif_fire & htif_pcr_req_addr == CSR_ADDR_TO_HOST & ~system_wen:
+                to_host.next = 0
+
+    return instances
```
Comment (jck) [inline riscv/csr_file.py]: instances is a function.

File: riscv/ctrl.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+    killed_DX = prev_killed_DX | kill_DX
+    ex_DX     = had_ex_DX | new_ex_DX
+
+    @always_comb
+    def DX_stage_comb():
+        ex_code_DX = ECODE_INST_ADDR_MISALIGNED
+        if had_ex_DX:
+            ex_code_DX = ECODE_INST_ADDR_MISALIGNED
+        elif illegal_instruction:
```
Comment (jck) [inline riscv/ctrl.py]: This doesn't seem right. You probably want to use .next instead of reassigning the variable. Does this module have a test?

Comment (meetshah1995) [inline riscv/ctrl.py]: Yes you are correct. It is in progress, the `controller` and the `csr_file` tests are the only tests remaining and we are working on it.

File: riscv/csr_file.py (hunk 2 above)
Comment (meetshah1995) [inline riscv/csr_file.py]: Yes you are correct. I missed the declaration. will make the change.

File: riscv/PC_mux.py (hunk 1 above)
Comment (meetshah1995) [inline riscv/PC_mux.py]: I didn't use it specifically, but incase the immediate generated overflows. The algol core also used modbv for such signals , that's why.

Comment (jck) [inline riscv/PC_mux.py]: I would suggest using intbv unless vscale explicitly does wrap around for a particular signal.

File: riscv/csr_constants.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+CSR_ADDR_WIDTH     = 12
+CSR_COUNTER_WIDTH  = 64
+
+CSR_ADDR_CYCLE     = int('C00',16)
+CSR_ADDR_TIME      = int('C01',16)
+CSR_ADDR_INSTRET   = int('C02',16)
+CSR_ADDR_CYCLEH    = int('C80',16)
+CSR_ADDR_TIMEH     = int('C81',16)
```
Comment (jck) [inline riscv/csr_constants.py]: Python supports hex literals like `0xC81`

Comment (meetshah1995) [inline riscv/csr_constants.py]: Oh, Dint know that, will make the change.

File: riscv/PC_mux.py (hunk 1 above)
Comment (meetshah1995) [inline riscv/PC_mux.py]: Ok, will make the change.

**Pull request merged**